go-controller/pkg/factory/factory.go

```diff
@@ -56,11 +57,40 @@ func (h *Handler) kill() error {
 	return nil
 }
 
+type eventKind int
+
+const (
+	addEvent eventKind = iota
+	updateEvent
+	deleteEvent
+)
+
+type event struct {
+	obj    interface{}
+	oldObj interface{}
+	kind   eventKind
+}
+
 type informer struct {
 	sync.Mutex
-	oType    reflect.Type
-	inf      cache.SharedIndexInformer
-	handlers map[uint64]*Handler
+	oType      reflect.Type
+	inf        cache.SharedIndexInformer
+	handlers   map[uint64]*Handler
+	eventsLock *sync.Mutex
```

Comment:
why do you need a separate lock here rather than just reusing the main mutex?

Reply:
@danwinship because we only need to protect the event array; we don't need to protect access to the event array from concurrent access of the handlers map or running handlers.

It turns out that we can use the main mutex, we just have to change it to an RWMutex and then RLock()/RUnlock() forEachQueuedHandler() and enqueueEvent() since evidently they contend with each other if they are using the same mutex and fail the testcases by blocking on each other too long or something.

Reply:
> we just have to change it to an RWMutex

oh... if that makes things ugly/complicated, I would have accepted "we can't use the main mutex because they'd contend with each other"